ui/src/dashboards/components/QueryEditor.tsx

```diff
@@ -0,0 +1,465 @@
+import React, {Component, MouseEvent} from 'react'
+import _ from 'lodash'
+import classnames from 'classnames'
+
+import TemplateDrawer from 'src/shared/components/TemplateDrawer'
+import QueryStatus from 'src/shared/components/QueryStatus'
+import {ErrorHandling} from 'src/shared/decorators/errors'
+import {Controlled as ReactCodeMirror, IInstance} from 'react-codemirror2'
+import {replace as replaceQueryTemplates} from 'src/shared/apis/query'
+
+import {Template, QueryConfig} from 'src/types'
+import {EditorChange} from 'codemirror'
+
+import {
+  MATCH_INCOMPLETE_TEMPLATES,
+  applyMasks,
+  insertTempVar,
+  unMask,
+} from 'src/tempVars/constants'
+
+interface State {
+  focused: boolean
+  queryText: string
+  editorValue: string
+  isTemplating: boolean
+  selectedTemplate: {
+    tempVar: string
+  }
+  isShowingTemplateValues: boolean
+  filteredTemplates: Template[]
+}
+
+interface Props {
+  query: string
+  onUpdate: (text: string) => void
+  config: QueryConfig
+  isInDataExplorer?: boolean
+  templates: Template[]
+}
+
+const CODE_MIRROR_OPTIONS = {
+  tabIndex: 1,
+  mode: 'influxQL',
+  readonly: false,
+  lineNumbers: false,
+  autoRefresh: true,
+  theme: 'influxql',
+  completeSingle: false,
+  lineWrapping: true,
+}
+
+const TEMPLATE_START = ':'.length
+
+const NOOP = () => {}
+const NULL_RESOLUTION = null
+const SHOW_TEMPLATES = 'Show Raw Query'
+const HIDE_TEMPLATES = 'Hide Raw Query'
+
+const BLURRED_EDITOR_STATE = {
+  focused: false,
+  isTemplating: false,
+  isShowingTemplateValues: false,
+}
+
+@ErrorHandling
+class QueryEditor extends Component<Props, State> {
```

Comment:
maybe this component is specifically an `InfluxQLEditor`? seems like this `QueryEditor` namespace is a hot market :)